keps/sig-node/4569-cgroup-v1-deprecation/README.md

```diff
@@ -0,0 +1,685 @@
+			return
+		}
+	}
+```
+Revised Code Snippet:
+```golang
+if m.memorySwapBehavior == kubelettypes.LimitedSwap {
+    if !isCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) && !SupressCgroupV1Warning() {
+      klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+      eventRecorder.Event(pod, v1.EventTypeWarning, "DeprecatedCgroupV1", fmt.Sprint("cgroup v1 support has been deprecated"))
+      swapConfigurationHelper.ConfigureNoSwap(lcr)
+      return
+    }
+}
+```
+
+Furthermore, code segments that default to cgroup v1 logic will be inverted to assume cgroup v2 as the default. This shift underscores the transition to cgroup v2 and ensures warning logs are consistently generated for cgroup v1 usage. e.g.
+
+Original Code Snippet:
+```golang
+	memLimitFile := "memory.limit_in_bytes"
+	if libcontainercgroups.IsCgroup2UnifiedMode() {
+		memLimitFile = "memory.max"
+	}
+```
+Revised Code Snippet:
+```golang
+	memLimitFile = "memory.max"
+	if libcontainercgroups.!IsCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) && !SupressCgroupV1Warning() {
+      klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+      eventRecorder.Event(pod, v1.EventTypeWarning, "DeprecatedCgroupV1", fmt.Sprint("cgroup v1 support has been deprecated"))
+      memLimitFile := "memory.limit_in_bytes"
+	}
+```
+
+### Feature Flag and Kubelet Option
+
+During the `alpha` stage of this feature's lifecycle, it's anticipated that users will require time to plan and execute their migrations to cgroup v2. Factors influencing the migration timeline include, but are not limited to, selecting compatible kernels and operating systems, planning for potential downtimes, and coordinating changes to minimize disruption during peak operational periods.
+
+To accommodate this, we propose introducing a kubelet flag, `--suppress-cgroupv1-deprecation-warning`, which is not set by default. This flag provides users the flexibility to suppress repetitive deprecation warning logs and events related to cgroup v1 while they are in the process of planning and executing their migration to cgroup v2. The introduction of this flag is intended to reduce log spam and operational noise, acknowledging the non-trivial nature of the migration for many users.
+
+#### Beta Stage Behavior Modification
+
+Upon reaching the `beta` stage, the kubelet's approach towards cgroup version enforcement will be adjusted to underline the importance of transitioning to cgroup v2. By default, kubelet will not start on systems that are still using cgroup v1. This change acts as a strong nudge for users to migrate to cgroup v2, ensuring they can utilize the enhanced features and capabilities of Kubernetes without compromise. Users who wish to continue with cgroup v1 need to explicitly opt-out by disabling the `DeprecatedCgroupV1` feature flag.
```

Comment:
This is a strong decision, dockershim deprecation took a while, the existing criteria in the KEP does not define any object metrics to indicate this is safe, we should have something as top X major distros and top Y cloud providers no longer supportcgroupsv1 or something like that, leaving this as a developer choice is risky and can break many users ...